src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java

```diff
@@ -2601,6 +2601,32 @@ private void readClassBuffer(ClassSymbol c) throws IOException {
     }
 
     public void readClassFile(ClassSymbol c) {
+        readClassFileInternal(c);
+        if (c.isValue()) {
```

Comment:
This approach leads to very compact code - but in javac-land it's a bit odd that we have to prune away members at this late stage. I wonder if some of the pruning could happen earlier (e.g. in TransValues?)

Reply:
Actually, this code is trying to remove V$ref.class from its owner (the package). The class finder would have from the class path found V.class and V$ref.class and added them to the known classes. When V.class is completed and we find that it is a value class, V.class has a "wrong" super type from javac pov. It need to read V$ref.class and pick up the "real" super type from there. In that process it needs to remove V$ref.class from the package owner.

Reply:
Actually I meant to comment on ClassWriter and clicked on the wrong thing!

Reply:
Actually we don't prune anything, We simply don't write the methods. But carry along the symbols as we must because there could be other files that are in the pipeline, (from same TU or otherwise)